src/main/java/com/icthh/xm/ms/entity/service/privileges/custom/EntityCustomPrivilegeService.java

```diff
@@ -0,0 +1,198 @@
+package com.icthh.xm.ms.entity.service.privileges.custom;
+
+import static com.icthh.xm.commons.config.client.repository.TenantConfigRepository.TENANT_NAME;
+import static com.icthh.xm.ms.entity.service.privileges.custom.CustomPrivilegesExtractor.DefaultPrivilegesValue.ENABLED;
+import static com.icthh.xm.ms.entity.service.privileges.custom.CustomPrivilegesExtractor.DefaultPrivilegesValue.NONE;
+import static java.util.Arrays.asList;
+import static java.util.Optional.ofNullable;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.icthh.xm.commons.config.client.repository.CommonConfigRepository;
+import com.icthh.xm.commons.config.domain.Configuration;
+import com.icthh.xm.commons.logging.aop.IgnoreLogginAspect;
+import com.icthh.xm.commons.permission.config.PermissionProperties;
+import com.icthh.xm.commons.permission.domain.Permission;
+import com.icthh.xm.commons.permission.domain.Role;
+import com.icthh.xm.commons.permission.service.RoleService;
+import com.icthh.xm.ms.entity.domain.spec.TypeSpec;
+import com.icthh.xm.ms.entity.service.privileges.custom.CustomPrivilegesExtractor.DefaultPrivilegesValue;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class EntityCustomPrivilegeService {
+
+    private static final String CUSTOMER_PRIVILEGES_PATH = "/config/tenants/{" + TENANT_NAME + "}/custom-privileges.yml";
+
+    private final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+
+    private final CommonConfigRepository commonConfigRepository;
+    private final PermissionProperties permissionProperties;
+    private final RoleService roleService;
+    private final List<CustomPrivilegesExtractor> privilegesExtractors;
+
+    @IgnoreLogginAspect
+    public void updateCustomPermission(Map<String, TypeSpec> specs, String tenantKey) {
+
+        String privilegesPath = resolvePathWithTenant(tenantKey, CUSTOMER_PRIVILEGES_PATH);
+        String specPath = permissionProperties.getPermissionsSpecPath();
+        String permissionsSpecPath = resolvePathWithTenant(tenantKey, specPath);
+
+        log.info("Get config from {} and {}", privilegesPath, permissionsSpecPath);
+        List<String> paths = asList(privilegesPath, permissionsSpecPath);
+        Map<String, Configuration> configs = commonConfigRepository.getConfig(null, paths);
+        configs = configs != null ? configs : new HashMap<>();
```

Comment:
Suggest invert this condition. See Robert C. Martin "Clean Code", "G29: Avoid Negative Conditionals" p. 302